### PR TITLE
Graceful shutdown

### LINKED ISSUE #15: Graceful shutdown
It might make sense to make RabbitMQ plugin Closeable, so that ktor would know how to dispose of it and automatically trigger on application shutdown:
```
class RabbitMQ(
    val configuration: RabbitMQConfiguration
): Closeable

    override fun close() {
        // Gracefull stop RabbitMQ instance here
    }
```

### PR DESCRIPTION
**Pull request by JUtupe**

### CODE REVIEW

File: src/main/kotlin/pl/jutupe/ktor_rabbitmq/RabbitMQ.kt [hunk 1]
```diff
@@ -35,6 +43,12 @@ class RabbitMQ(
     fun <T> serialize(body: T): ByteArray =
         configuration.serializeBlock.invoke(body!!)
 
+    override fun close() {
```
Comment (kibertoad) [inline src/main/kotlin/pl/jutupe/ktor_rabbitmq/RabbitMQ.kt:46]: shouldn't it close channel and connection?

Comment (JUtupe) [reply src/main/kotlin/pl/jutupe/ktor_rabbitmq/RabbitMQ.kt:46]: As per Rabbit client's ConnectionFactory:

> Set the executor to use for connection shutdown. All connections that use this executor share it

I think that'll make this plugin's graceful shutdown to work as expected.

File: src/main/kotlin/pl/jutupe/ktor_rabbitmq/RabbitMQ.kt [hunk 2]
```diff
@@ -35,6 +43,12 @@ class RabbitMQ(
...
+        logger?.info("RabbitMQ stopping")
+        shutdownExecutor.awaitTermination(configuration.shutdownTimeout, TimeUnit.MILLISECONDS)
```
Comment (kibertoad) [inline src/main/kotlin/pl/jutupe/ktor_rabbitmq/RabbitMQ.kt:48]: you are not invoking shutdown first, though. See https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html
from my understanding one should first shutdown, and then await termination.

Comment (kibertoad) [reply src/main/kotlin/pl/jutupe/ktor_rabbitmq/RabbitMQ.kt:48]: we should revise this pr

Comment (JUtupe) [reply src/main/kotlin/pl/jutupe/ktor_rabbitmq/RabbitMQ.kt:48]: Yeah, sure. I'll give it another try on ktor 2.0

**Comment by JUtupe** [comment]
I'll try to implement it again using ktor 2.0

**Pull request closed without merging**